src/herder/HerderSCPDriver.cpp

```diff
@@ -324,39 +342,29 @@ HerderSCPDriver::validateValueHelper(uint64_t slotIndex, StellarValue const& b,
         return SCPDriver::kInvalidValue;
     }
 
-    if ((!nomination || lcl.ledgerVersion < 11) &&
-        b.ext.v() != STELLAR_VALUE_BASIC)
-    {
-        // ballot protocol or
-        // pre version 11 only supports BASIC
-        CLOG(TRACE, "Herder")
-            << "HerderSCPDriver::validateValue"
-            << " i: " << slotIndex << " invalid value type - expected BASIC";
-        return SCPDriver::kInvalidValue;
-    }
-    if (nomination &&
-        (lcl.ledgerVersion >= 11 && b.ext.v() != STELLAR_VALUE_SIGNED))
-    {
-        // v11 and above use SIGNED for nomination
-        CLOG(TRACE, "Herder")
-            << "HerderSCPDriver::validateValue"
-            << " i: " << slotIndex << " invalid value type - expected SIGNED";
-        return SCPDriver::kInvalidValue;
-    }
-
     Hash const& txSetHash = b.txSetHash;
     TxSetFramePtr txSet = mPendingEnvelopes.getTxSet(txSetHash);
 
     SCPDriver::ValidationLevel res;
 
+    // In the presence of the "closetime change" for protocol issue 622,
+    // the close time to validate against is not the last ledger close
+    // time, but the proposed close time (which we have validated above,
+    // in checkCloseTime(), is greater than the last ledger close time).
+    if (b.closeTime <= lastCloseTime)
+    {
+        throw std::logic_error("Invalid proposed ledger close time");
+    }
+
     if (!txSet)
     {
         CLOG(ERROR, "Herder") << "validateValue i:" << slotIndex
                               << " unknown txSet " << hexAbbrev(txSetHash);
 
         res = SCPDriver::kInvalidValue;
     }
-    else if (!txSet->checkValid(mApp, 0))
+    else if (!txSet->checkValid(mApp, closeTimeOffsetToTrimOrValidateAgainst(
```

Comment:
I don't think this is what CAP0034 says: we expect txsets to all be valid for the new close time.

This change implies that we expect them to be valid from `lcl.ct` up to `new_ct` (inclusive) which is a lot more restrictive.

Reply:
I had failed to glean that that meant to check over a range and not an offset.  I guess the difference is in whether it returns `txTOO_EARLY` or not; is that right, or are there any other potential differences?

Reply:
And actually, don't we check a transaction for validity as soon as we receive it, using the last ledger header close time (in particular to check for `txTOO_EARLY`), in `TransactionQueue::canAdd()`?  And when a ledger closes, we trim transactions against the new last ledger close time, in `updateTransactionQueue()`, correct?

Consequently, won't we already have validated a transaction against the last ledger header close time before we reach the quoted check, so that checking for `txTOO_EARLY` against the last ledger header close time is harmless (redundant), and won't happen, and the only semantic effect of the change is the intended CAP-34 effect of checking for `txTOO_LATE` against a later `closeTime`?

Reply:
> And actually, don't we check a transaction for validity as soon as we receive it, using the last ledger header close time (in particular to check for `txTOO_EARLY`), in `TransactionQueue::canAdd()`? And when a ledger closes, we trim transactions against the new last ledger close time, in `updateTransactionQueue()`, correct?

Upon even further thought, I think that my second sentence, about `updateTransactionQueue()`, is unnecessary.  If a transaction in the queue were going to be `txTOO_EARLY` at the time of a ledger close, then it would have been `txTOO_EARLY` at the time of any _previous_ ledger close, too, so it would have been rejected before even entering the queue, in `TransactionQueue::canAdd()`.  So the validation in `TransactionQueue::canAdd()` has, I believe, already ensured that any transaction that makes it to `validateValueHelper()` (above) or `triggerNextLedger()` (below) _is_ valid with respect to `lcl.ct`.  Validating again against `lcl.ct` (as well as `new_ct`) at those points is therefore not any more restrictive than validating against that whole range.

Reply:
> Validating again against lcl.ct (as well as new_ct) at those points is therefore not any more restrictive than validating against that whole range.

Except that, on yet _more_ consideration, I realize that that's only assuming that you're well-behaved!  When we validate a `StellarValue` in `triggerNextLedger()` or `validateValueHelper()`, it might have come from an ill-behaved node.  So the question is, what's the correct behavior if the `StellarValue` came from an ill-behaved node which had not validated against `lcl.ct` when it accepted the transaction into its queue -- to check `txTOO_EARLY` the way my code does, and as the code before CAP-0034 already did, against `lcl.ct`, or the way you suggest, against `new_ct`?

I have not, so far, been able to think of an attack that a bad validator could use to take advantage of the weaker validation that you suggest, but in general it seems to me that if the runtime cost is the same, then we should validate as strongly as we believe we should be able to.  Put another way:  if some other node did something that we ourselves _wouldn't_ have done in the same situation, then I think we probably shouldn't trust it.

Reply:
You're confusing what is defined at the application protocol level from the rest:
* at the protocol level, we have very strict set of validity checks. Basically anything that happens during consensus. Those rules have to be as minimal as possible as to avoid tying up changes in overlay to protocol changes.
* outside of the protocol, we have other things that we can do but that are not mandated at the network level. Typically those things are there to make our life easier and can change at any time, or just the result that individual nodes on the network are not trusted. This includes:
    * whatever heuristic we want to use at the overlay layer.
    * the transaction set composition (ie, how to fit as many good transactions in a transaction set)

With that said: the concept of validity range is *not* part of the application level protocol today, I don't see a reason to add this as it creates more problems than it solves (and a PR is not the place to discuss this anyways).

Reply:
> With that said: the concept of validity range is _not_ part of the application level protocol today, I don't see a reason to add this as it creates more problems than it solves (and a PR is not the place to discuss this anyways).

I don't understand why you're saying that "validity range" isn't part of the protocol today. Today, a `StellarValue` `v` is valid (as far as `closeTime` is concerned) if and only if the following hold:

```
max(map(minTime,v.txs)) <= lcl.closeTime
min(map(maxTime,v.txs)) >= lcl.closeTime
```

In other words, `lcl.closeTime` already has a validity range, namely `[max(map(minTime,v.txs))]`.

After my change, a `StellarValue` `v` is valid (as far as `closeTime` is concerned) if and only if the following hold:

```
max(map(minTime,txs(v))) <= lcl.closeTime
min(map(maxTime,txs(v)) >= v.closeTime
```

I haven't changed the form of the inequalities; I've just made the second inequality stricter (it's definitely stricter as opposed to equivalent or weaker, because we also validate `v.closeTime > lcl.closeTime`).

Or, if you look at it from the perspective of the individual transactions rather than the `StellarValue`, then today, a given transaction `tx` is valid (as far as `closeTime` is concerned) if and only if the following hold:

```
minTime(tx) <= lcl.closeTime
maxTime(tx) >= lcl.closeTime
```

So `minTime(tx)` already had a validity range, namely `(-inf,lcl.closeTime])`, and `maxTime(tx)` already had a validity range, namely `[lcl.closeTime,+inf)`.

After my change, a given transaction `tx` is valid in a `StellarValue` `v` (as far as `closeTime` is concerned) if and only if the following hold:

```
minTime(tx) <= lcl.closeTime
maxTime(tx) >= v.closeTime
```

So in my code, `minTime(tx)`'s validity range remains the same, and `maxTime(tx)`'s validity gets smaller on the left, becoming `[v.closeTime,+inf)`.

How is it that validity ranges are not there already?  It looks to me as though they're already there, and I'll I've done is change the location of one of the boundaries (which I thought is exactly what CAP-34 said I should do).

Reply:
Nicolas has explained that he _does_ want to give us the flexibility to relax the flooding as well in the future, so it would be best to weaken the test for `txTOO_EARLY` to the precise degree that we can, namely, against the `closeTime` in the `StellarValue` itself, like the test for `txTOO_LATE` -- though the latter is a strengthening.  So the overall change that we want is not (as I have been thinking) a strengthening of validation, but a strengthening of one test and a weakening of another, with the overall result being neither a weakening nor a strengthening, but a _shift_, of the whole mechanism of `minTime`/`maxTime`/`closeTime` validation from `lcl.closeTime` to `v.closeTime`.

Reply:
(I'll make that semantic change after addressing other code review comments.)

Reply:
(I'll also update the CAP to reflect that.)

Reply:
Alright, I have made the big semantic change in commit 14a693a04985274b003047da09196554b82cb3bc, by extending the PR #2608 code to include a `lowerBoundCloseTimeOffset` as well as the `upperBoundCloseTimeOffset` which that PR added.  That allows my new CAP-34 code to use the `lowerBoundCloseTimeOffset` to specify which close time to validate `minTime` against, as it was already using `upperBoundCloseTimeOffset` to specify which close time to validate `maxTime` against.  (In the two key places where the semantics change -- `HerderImpl::triggerNextLedger()` and `HerderSCPDriver::validateValueHelper()` -- the CAP-34 code passes the same value for both offsets (the difference between the proposed next `closeTime` and the last ledger close time, meaning that both `minTime` and `maxTime` will be validated against the proposed next `closeTime`).)